Refactored pkg/scheduler/utils to podinfo and heap packages.

hprateek43 opened the pull request:
**What type of PR is this?**
kind cleanup

**What this PR does / why we need it:**
#71863 is an active open request for scheduler cleanup. I have picked up that request and moved the generic util package into more meaningful internal packages podinfo and heap. Necessary changes in the scheduler and kubelet eviction packages have been made to support those changes.

Fixes #71863

**Special notes for your reviewer**:
The previous PR used schedulerinternalpodinfo as the package reference.The name was found to be too long by members in the previous thread and has been reduced to just podinfo. Let me know if we need any refactoring to rename this to something more meaningful. Also the package is in /pkg/scheduler/internal/podinfo and /pkg/scheduler/internal/heap, so the word internal in path can be a potential break. Comment if we can move them to more meaningful areas.

```release-note
NONE
```

hprateek43 commented:
@misterikkit Can you review this PR and let me know the changes ?

hprateek43 commented:
@lavalamp @derekwaynecarr Can you review this request?

lavalamp commented:
You should get a scheduler-specific review, e.g. @bsalamat or @misterikkit.

I'm not sure why the go.sum files all changed; also, the commits seem arranged sub-optimally (consider using fetch & rebase rather than merge?).

liggitt commented on go.mod:
```diff
@@ -157,6 +157,7 @@ require (
 	github.com/storageos/go-api v0.0.0-20180912212459-343b3eff91fc
 	github.com/stretchr/testify v1.2.2
 	github.com/syndtr/gocapability v0.0.0-20160928074757-e7cb7fa329f4 // indirect
+	github.com/tools/godep v0.0.0-20180126220526-ce0bfadeb516 // indirect
```
drop this commit

misterikkit replied:
Yeah, this shouldn't be included in the PR. Is the change needed at all?

hprateek43 replied:
@misterikkit @liggitt Squashed and rebased. Dropped commits as per recommendation

liggitt commented on staging/src/k8s.io/api/go.sum:
```diff
@@ -30,11 +30,14 @@ github.com/spf13/pflag v1.0.1 h1:aCvUg6QPl3ibpQUxyLkrEkCHtPqYJL4x9AuhqVqFis4=
 github.com/spf13/pflag v1.0.1/go.mod h1:DYY7MBk1bdzusC3SYhjObp+wFpr4gzcvqqNjLnInEg4=
 github.com/stretchr/testify v1.2.2 h1:bSDNvY7ZPG5RlJ8otE/7V6gMiyenm9RtJ7IUVIAoJ1w=
 github.com/stretchr/testify v1.2.2/go.mod h1:a8OnRcib4nhh0OaRAV+Yts87kKdq0PP7pXfy6kDkUVs=
+golang.org/x/net v0.0.0-20180906233101-161cd47e91fd/go.mod h1:mL1N/T3taQHkDXs73rZJwtUhF3w3ftmwwsq0BUmARs4=
```
drop this commit as well

hprateek43 replied:
@liggitt Dropped

liggitt commented:
there shouldn't be any go.{sum,mod} changes required from what I can see. I'd suggest squashing/rebasing on master and dropping the changes to those files

dims commented:
/unassign

hprateek43 commented:
@liggitt Will do so. Should I amend the commit of this PR or open a new PR?

hprateek43 commented:
> there shouldn't be any go.{sum,mod} changes required from what I can see. I'd suggest squashing/rebasing on master and dropping the changes to those files

I had create a new module for podinfo and heap in /pkg/scheduler/internal. Are the changes in go.{mod|sum} due to that?

misterikkit commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -19,6 +19,7 @@ package predicates
 import (
 	"context"
 	"fmt"
+	"k8s.io/kubernetes/pkg/scheduler/internal/podinfo"
```
Please move this import down to the other `k8s.io` imports. The preferred style is usually,
```
<standard library imports>
<blank line>
<everything else>
```

When I'm writing imports, I like to add one extra level of organization:
```
<standard library imports>
<blank line>
<libraries not part of scheduler>
<blank line>
<scheduler libraries>
```

misterikkit replied:
(and please make the same fix in the other files as well.)

hprateek43 replied:
This happened as a result of auto refactoring by GoLand. I have squashed and rebased the commits. Please review the same.

misterikkit replied:
Looks like the import is still in the top section. Can you override GoLand's placement? (if you manually move it, does GoLand put it back?)

misterikkit commented on pkg/scheduler/internal/heap/heap.go:
```diff
@@ -56,7 +57,7 @@ type heapData struct {
 	keyFunc KeyFunc
 	// lessFunc is used to compare two objects in the heap.
-	lessFunc LessFunc
+	lessFunc podinfo.LessFunc
```
It feels to me like `LessFunc` belongs in this `heap` package. Is it still used anywhere within the `podinfo` package?

misterikkit replied:
It looks like `LessFunc` is used by both `SortableList` and the heap package. If I had a lot of time on my hands, there are ways I would want to reorganize/refactor this. But I think it's good enough as-is.

ravisantoshgudimetla replied:
I agree, perhaps a TODO within the codebase?

misterikkit commented on pkg/scheduler/internal/podinfo/clock.go:
```diff
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions and
 */
 
-package util
+package podinfo
```
I forgot we had clock stuff in here. I should probably file a separate issue to delete that. :smile:

hprateek43 replied:
@misterikkit We should delete the clock stuff or should we migrate it into another package ?

misterikkit replied:
For this PR, I'd like to keep the clock stuff in the podinfo package.

After this PR, I want to delete it since there is already clock functionality in https://godoc.org/k8s.io/apimachinery/pkg/util/clock.

misterikkit commented on pkg/scheduler/internal/podinfo/utils.go:
```diff
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions and
 */
 
-package util
+package podinfo
```
Please rename this file to `podinfo.go`.

hprateek43 replied:
utils.go -> podinfo.go
utils_test.go -> podinfo_test.go

Is this the expected rename ?

misterikkit replied:
yes, thanks

misterikkit reviewed (commented):
Thanks for sending this cleanup! Except for a few comments, this looks good.

I would prefer if the PR title and the git commit message described the changes being made rather than the issue being fixed.

hprateek43 commented:
@liggitt @misterikkit Do the changes look more aligned now? I have omitted the rename change for now. If the changes look good let us run a test on these, and then I will rebase the commits to introduce the file rename. Just looking to check if something is breaking or not.

liggitt commented:
changes are now contained to the scheduler package, will defer to @kubernetes/sig-scheduling-pr-reviews for the rest

hprateek43 commented:
@misterikkit Modified the PR title to actual refactoring being done. Let me know if a better PR title can be used.

misterikkit commented on pkg/kubelet/eviction/helpers.go:
```diff
@@ -97,6 +97,18 @@ func validSignal(signal evictionapi.Signal) bool {
 	return found
 }
 
+// getPodPriority return priority of the given pod.
```
s/return/returns/

misterikkit reviewed (changes requested):
We will still need approval for pkg/controller and pkg/kubelet, but let's get scheduler approval first.

Please go ahead with the file renaming and squash commits.

hprateek43 commented:
@misterikkit Went back to basics with vim and changed the import placement locations. I have also renamed the files in the /scheduler/internal/podinfo section.

hprateek43 commented:
@bowei @andrewsykim Can you review this PR?

misterikkit commented:
/lgtm

looks like you will need a rebase to resolve merge conflicts.

hprateek43 commented:
@misterikkit I am awaiting the test results before I rebase. Will I get them as it seems to take forever and the LGTM label has been removed?

hprateek43 commented:
/test pull-kubernetes-verify

misterikkit commented:
/ok-to-test

The tests that matter can be run on your local machine without having to
wait forever. Either with `make test` or `bazel test`

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/testing.md

*From:*Kubernetes Prow Robot <notifications@github.com>
*Date:*Fri, May 3, 2019, 7:01 AM
*To:*kubernetes/kubernetes
*Cc:*Jonathan Basseri, Mention

@hprateek43 <https://github.com/hprateek43>: Cannot trigger testing until a
> trusted user reviews the PR and leaves an /ok-to-test message.
>
> In response to this
> <https://github.com/kubernetes/kubernetes/pull/76990#issuecomment-489104822>
> :
>
> /test pull-kubernetes-verify
>
> Instructions for interacting with me using PR comments are available here
> <https://git.k8s.io/community/contributors/guide/pull-requests.md>. If
> you have questions or suggestions related to my behavior, please file an
> issue against the kubernetes/test-infra
> <https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:>
> repository.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/76990#issuecomment-489104903>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AANWVKVKXTJLOUCCGWF6IELPTRAUNANCNFSM4HIA42XQ>
> .
>

misterikkit commented:
I chatted with @hprateek43 on May 6 about resolving the merge here. Other reviewers feel free to wait until we get that resolved.

hprateek43 commented:
@misterikkit I was out on travel recently. Will be picking up this today and sending the commit. Just need to handle the test cases now.

hprateek43 commented:
@misterikkit I would probably need you to review the commits and resolve the conflicting files in the commits. I think only users with write access can perform that.

hprateek43 commented:
@misterikkit Can you review and add lgtm ?

hprateek43 commented:
@bowei @andrewsykim Can you review this PR?

ahg-g commented on pkg/kubelet/eviction/helpers.go:
```diff
@@ -18,20 +18,19 @@ package eviction
+	"sort"
+	"strconv"
+	"strings"
+	"time"
```
can you please keep the order as it was before? those imports should be at the top.

misterikkit replied:
+1
I'm surprised lint didn't complain about this.

ahg-g commented on pkg/kubelet/eviction/helpers.go:
```diff
@@ -91,6 +90,16 @@ func init() {
 	signalToResource[evictionapi.SignalPIDAvailable] = resourcePids
 }
 
+func getPodPriority(pod *v1.Pod) int32 {
```
we now have two implementations of this, I think we should have a single implementation under pkg/api/pod/util.go

misterikkit replied:
This was discussed in a previous (closed) PR, but for something small like this, having two copies is not the end of the world. I support moving it to a neutral library (neither kubelete nor scheduler) but I don't think that should be part of this PR.

ahg-g replied:
Not doing it in this PR is fine, but we should move it eventually, even small things like this can diverge in semantics without noticing.

ravisantoshgudimetla replied:
Or add a todo

ahg-g reviewed (commented):
the packages are still named util, we should change it to heap and podinfo.

misterikkit commented on pkg/controller/.import-restrictions:
```diff
@@ -252,7 +252,7 @@
         "k8s.io/kubernetes/pkg/master/ports",
         "k8s.io/kubernetes/pkg/scheduler/api",
-        "k8s.io/kubernetes/pkg/scheduler/util",
+        "k8s.io/kubernetes/pkg/scheduler/internal",
```
Is this line needed? Was it auto-generated?

hprateek43 replied:
I think it is autogenerated. Since I refactored util to internal, GoLand changed the dependency there. I am not sure if it is needed, but will check if it is required.

hprateek43 replied:
We need this change. Just went through the dependencies.

misterikkit reviewed (commented):
This just needs a couple tweaks before LGTM.

Some of the moved files still have `package util` at the top. Those need to be renamed.

hprateek43 commented:
/retest

hprateek43 commented:
/retest

hprateek43 commented:
/retest pull-kubernetes-bazel-test

hprateek43 commented:
@misterikkit Can you review the changes now?

hprateek43 commented:
@andrewsykim @bowei @misterikkit Can you review and give a /lgtm ?

hprateek43 commented:
@deads2k @sig-scheduling-maintainers @Random-Liu Can you approve this PR?

deads2k commented on pkg/controller/.import-restrictions:
```diff
@@ -253,7 +253,7 @@
         "k8s.io/kubernetes/pkg/master/ports",
         "k8s.io/kubernetes/pkg/scheduler/api",
-        "k8s.io/kubernetes/pkg/scheduler/util",
+        "k8s.io/kubernetes/pkg/scheduler/internal",
```
This breaks a logical dependency contract.  The controller and scheduler are distinct.  Having the controller depend on an internal package in the scheduler means the two can never be separated.  This is a bad outcome.  We should not make this change.

deads2k replied:
> This breaks a logical dependency contract. The controller and scheduler are distinct. Having the controller depend on an internal package in the scheduler means the two can never be separated. This is a bad outcome. We should not make this change.

hmmm.... Tests passed.  I'm wondering if this is a transitive dep and this file just lists all possible transitives you can rely upon.

hprateek43 replied:
I think we can remove this now post recent rebases. Will force push an update if local builds go through

ravisantoshgudimetla replied:
@deads2k - You're right, `k8s.io/kubernetes/pkg/controller/daemon` has a dependency on `pkg/scheduler/algorithm/predicates which in turn has a dependency on `k8s.io/kubernetes/pkg/scheduler/internal`(or the util before renaming). So, we should be ok with this but we need to see if we can refactor the predicates into a more neutral place that so both scheduler and DS controller can reuse them so that we can remove the dependecy

ravisantoshgudimetla replied:
/cc @bsalamat

liggitt replied:
> we should be ok with this but we need to see if we can refactor the predicates into a more neutral place that so both scheduler and DS controller can reuse them so that we can remove the dependency

we should do that *before* moving them into internal, rather than moving in a direction we don't want, then recovering

misterikkit replied:
FYI I've discussed `.import-restrictions` with @lavalamp on past PRs and it's true that they list transitive dependencies. My opinion is that this is a quirk of the tool, and transitive dependencies on `pkg/scheduler/internal` are no worse than direct dependencies on `pkg/scheduler`

There are various k8s components that import bits of the scheduler code base, somewhat ad hoc. Moving code into our internal folder is a way to draw a line in the sand and prevent more packages from leaking into our supported API surface.

I also hope that getting explicit about which libraries are reasonable to use in other components will lead to better APIs and separation of concerns down the road.

deads2k commented:
See https://github.com/kubernetes/kubernetes/pull/76990/files#r294937633 . The controller-manager must not rely on an internal scheduler package.  We should not pursue this pull if it means that internal packages are leaking into other processes.  If those those processes need their dependencies fixed, do that first, then move the package.

/hold

deads2k commented:
/assign @ravisantoshgudimetla 

have a look to see if you know of anyone already depending on these packages before they move to internal.

hprateek43 commented:
/retest pull-kubernetes-integration

hprateek43 commented:
/test pull-kubernetes-integration

hprateek43 commented:
/test pull-kubernetes-kubemark-e2e-gce-big

ravisantoshgudimetla reviewed (commented):
@hprateek43 - Thanks for the PR. I am fine with these changes, as long as you add some TODOs and create follow-up PRs on them
/lgtm

/hold

I want to give sometime to @deads2k and @misterikkit if they have any additional comments.

deads2k commented:
Ok, since the internal dep is a transitive, I'll 

/approve

hprateek43 commented:
/assign @derekwaynecarr @ravisantoshgudimetla

hprateek43 commented:
@misterikkit @bsalamat Can you review and approve this merge ?

hprateek43 commented:
@deads2k Can you review this now?

misterikkit commented on pkg/kubelet/eviction/helpers.go:
```diff
@@ -513,8 +523,12 @@ func (ms *multiSorter) Less(i, j int) bool {
 func priority(p1, p2 *v1.Pod) int {
-	priority1 := schedulerutils.GetPodPriority(p1)
-	priority2 := schedulerutils.GetPodPriority(p2)
+	if !utilfeature.DefaultFeatureGate.Enabled(features.PodPriority) {
```
Did this check come in because of a rebase?

hprateek43 replied:
Yes. The recent rebase added this definition.

misterikkit replied:
Hmm, if this check was added and you did a rebase, I would not expect it to show up as an added line in the diff.

It looks like what actually happened is that this check was removed in https://github.com/kubernetes/kubernetes/pull/79262 but is getting re-added in this PR. Could you remove it?

hprateek43 replied:
I have removed the feature gate. Please review now.

misterikkit commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -39,8 +39,9 @@ import (
 	"k8s.io/kubernetes/pkg/scheduler/algorithm/predicates"
 	priorityutil "k8s.io/kubernetes/pkg/scheduler/algorithm/priorities/util"
 	framework "k8s.io/kubernetes/pkg/scheduler/framework/v1alpha1"
+	heap "k8s.io/kubernetes/pkg/scheduler/internal/heap"
```
these import aliases are redundant. If your IDE added them, then I don't want to spend effort on it. If you're manually adding them, I think we should remove it.

hprateek43 replied:
My IDE added this redundant import. Do let me know if I should refactor it?

misterikkit replied:
If your IDE added them, I don't want you (or anyone) to waste time manually fixing it. I think `goimports` somewhat recently changed its behavior around adding aliases. Ultimately, it's not doing any harm.

misterikkit replied:
If your IDE added it, I don't want you (or anyone) wasting time manually fixing it. I think that `goimports` changed how it does import aliases a while back and it's causing things like this. In the end, it is not causing harm.

hprateek43 replied:
@misterikkit Ravi told me that he was awaiting some changes to be incorporated from @liggitt 's review. Do you have any suggestions of changes to be made into this PR ?

misterikkit replied:
For import aliases, no changes are needed. For my other comment about the feature gate condition, that needs to change.

hprateek43 replied:
I have removed that condition now.

misterikkit reviewed (commented):
Looks like some of the files were moved from `pkg/scheduler/util` without changing the package name. I'm a little surprised the code compiles as-is.

misterikkit commented:
/lgtm
I am happy with this. Thanks! :smile:

Need someone with approve powers to review.
@bsalamat @dashpole could you take a look?

hprateek43 commented:
/retest

misterikkit reviewed (commented):
/lgtm

misterikkit commented:
/lgtm cancel

Looks like build is failing after the rebase. @hprateek43 can you fix it up?

hprateek43 commented:
> /lgtm cancel
> 
> Looks like build is failing after the rebase. @hprateek43 can you fix it up?

I am working on this. Seems like some BUILD files have changed post this rebase

hprateek43 committed:
Refactored pkg/scheduler/utils to podinfo, channels and heap packages.

hprateek43 commented:
/retest

hprateek43 commented:
/retest

hprateek43 commented:
/retest

hprateek43 commented:
/retest

hprateek43 commented:
/retest pull-kubernetes-integration

hprateek43 commented:
/retest

hprateek43 committed:
Refactored /pkg/kubelet

hprateek43 committed:
Updated /pkg/controler import restrictions

hprateek43 commented:
/retest

hprateek43 commented:
@misterikkit pull-kubernetes-integration has been a flaky test for me as most of the time some pod is getting timed out. Should I report this behaviour somewhere?

misterikkit reviewed (commented):
/lgtm

hprateek43 commented:
@bowei @andrewsykim @bsalamat Can you review and approve this PR?

hprateek43 commented:
@derekwaynecarr @vishh @dchen1107 @dashpole @sjenning Can one of you review this PR and give an approval ?

derekwaynecarr commented:
kubelet changes lgtm.

/approve
/lgtm

hprateek43 commented:
/assign @ravisantoshgudimetla

draveness commented on pkg/scheduler/internal/channels/error_channel.go:
```diff
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions and
 */
 
-package util
+package channels
```
`channels` may not be a good name for package, `channel` would be better as a package name here.

hprateek43 replied:
@draveness @alculquicondor What package name would your suggest to go with?

draveness replied:
I prefer `errchannel`.New` to `channel.NewErrorChannel`

alculquicondor replied:
being more specific, as @draveness suggested, reads well too.

draveness commented on pkg/scheduler/internal/heap/heap.go:
```diff
@@ -25,6 +25,7 @@ import (
 	"fmt"
 
 	"k8s.io/client-go/tools/cache"
+	podinfo "k8s.io/kubernetes/pkg/scheduler/internal/podinfo"
```
does goimports add this automatically or it's a typo.

draveness commented on pkg/scheduler/framework/v1alpha1/framework.go:
```diff
@@ -28,7 +28,7 @@ import (
 	"k8s.io/kubernetes/pkg/scheduler/apis/config"
 	"k8s.io/kubernetes/pkg/scheduler/internal/cache"
-	schedutil "k8s.io/kubernetes/pkg/scheduler/util"
+	channels "k8s.io/kubernetes/pkg/scheduler/internal/channels"
```
I think goimports only add alias when the package name is `v1`, it's the first time I saw this and looks weird.

Huang-Wei replied:
+1, that's also my understanding. Please remove this unnecessary alias.

draveness commented on pkg/scheduler/algorithm/priorities/interpod_affinity.go:
```diff
@@ -20,16 +20,16 @@ import (
 	"k8s.io/kubernetes/pkg/scheduler/algorithm/predicates"
 	priorityutil "k8s.io/kubernetes/pkg/scheduler/algorithm/priorities/util"
 	schedulerapi "k8s.io/kubernetes/pkg/scheduler/api"
+	schedutil "k8s.io/kubernetes/pkg/scheduler/internal/channels"
```
what do you think of keeping the original name of the package? rename it to `schedutil` does not sound like a good idea.

alculquicondor replied:
+1

draveness commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -22,17 +22,19 @@ import (
 	"math"
 	"sync"
 
+	"k8s.io/kubernetes/pkg/scheduler/internal/channels"
```
This should be grouped with imports in L29-L37

alculquicondor commented on pkg/scheduler/core/extender_test.go:
```diff
@@ -198,7 +198,7 @@ func (f *FakeExtender) selectVictimsOnNodeByExtender(
 	nodeInfoCopy := f.cachedNodeNameToInfo[node.GetName()].Clone()
 
-	potentialVictims := util.SortableList{CompFunc: util.MoreImportantPod}
+	potentialVictims := podinfo.SortableList{CompFunc: podinfo.MoreImportantPod}
```
It feels like SortableList shouldn't belong to the podinfo package.

hprateek43 replied:
@alculquicondor Can you suggest a package to move it into?

alculquicondor replied:
what about `internal/collections` or more specific `internal/sortablelist`

alculquicondor reviewed (commented):
My general feeling is that this PR can be broken up into at least 2.

For example:
PR1: move `Get*(v1.Pod)` into a new podinfo package
PR2: move heap into its own package
PR3: move SortableList into its own package.

arguably, heap and sortable list can live in the same package named. Arguably, heap and sortable list could be in the same package named "datastructures" or "containers".

hprateek43 commented:
@alculquicondor @draveness Can you suggest the right PR titles to split this PR?

alculquicondor commented:
Your call. If something is very off, I will request a change.

alejandrox1 commented:
/remove-area e2e-test-framework
/remove-sig testing

neolit123 commented:
/remove-area kubeadm

danwinship commented:
/remove-sig network

alculquicondor commented:
/close
This is quite stale now.
Perhaps the pod utils can be moved to a new package once #90039 lands
